core/src/test/java/org/apache/calcite/rex/RexProgramTest.java

```diff
@@ -931,15 +931,17 @@ private void checkExponentialCnf(int n) {
     checkSimplify(rexBuilder.makeCall(SqlStdOperatorTable.IS_NOT_NULL, aRef),
         "true");
 
-    // condition, and the inverse - nothing to do due to null values
-    checkSimplify2(and(le(hRef, literal(1)), gt(hRef, literal(1))),
-        "AND(<=(?0.h, 1), >(?0.h, 1))",
+    // condition, and the inverse
+    checkSimplify3(and(le(hRef, literal(1)), gt(hRef, literal(1))),
+        "<>(?0.h, ?0.h)",
+        "false",
         "false");
```

Comment:
It looks like the intention in the test was related to `null` somehow, however, `hRef` is `INTEGER NOT NULL`, so this expression is always `false`.

I thought Sarg was supposed to handle `x<>x` for non-nullable `x` and transform it to `true`

Reply:
You are correct. Ideally it would simplify to "false" even under unknown-as-unknown.

Can you please log a bug? (Not a sub-task, and not marked 'fix in 1.27'.)

I checked, and that case does not invoke Sarg logic, because the expression is deemed 'too simple'. Whoever fixes the bug can revisit that logic.